File: IO.Astrodynamics/Math/Plane.cpp
/*
 Copyright (c) 2023. Sylvain Guillet (sylvain.guillet@tutamail.com)
 */

//
// Created by spacer on 8/17/23.
//

#include <algorithm>
#include <Plane.h>
#include <cmath>

IO::Astrodynamics::Math::Plane::Plane(IO::Astrodynamics::Math::Vector3D normal, double distance) : m_normal{normal}, m_distance{distance}
{

}

double IO::Astrodynamics::Math::Plane::GetAngle(const IO::Astrodynamics::Math::Plane &plane) const
{
    return std::acos(m_normal.DotProduct(plane.m_normal) / (m_normal.Magnitude() * plane.m_normal.Magnitude()));
}

double IO::Astrodynamics::Math::Plane::GetAngle(const IO::Astrodynamics::Math::Vector3D &vector3D) const
{
    return std::asin(m_normal.DotProduct(vector3D) / (m_normal.Magnitude() * vector3D.Magnitude()));
}

File: IO.Astrodynamics.Net/IO.Astrodynamics/Math/Plane.cs
namespace IO.Astrodynamics.Math;

public record struct Plane
{
    public Vector3 Normal { get; }
    public double Distance { get; }

    public static Plane X { get; } = new Plane(Vector3.VectorX);
    public static Plane Y { get; } = new Plane(Vector3.VectorY);
    public static Plane Z { get; } = new Plane(Vector3.VectorZ);

    public Plane(Vector3 normal) : this(normal, 0.0)
    {
    }

    public Plane(Vector3 normal, double distance)
    {
        Normal = normal;
        Distance = distance;
    }

    public double GetAngle(Plane plane)
    {
        return System.Math.Acos((Normal * plane.Normal) / (Normal.Magnitude() * plane.Normal.Magnitude()));
    }

    public double GetAngle(Vector3 vector)
    {
        return System.Math.Asin((Normal * vector) / (Normal.Magnitude() * vector.Magnitude()));
    }
}